drivers/adc_ng/include/adc_ng_internal.h

```diff
@@ -0,0 +1,289 @@
+
+/**
+ * @brief   Use this value in @ref adc_ng_driver_t::fixed_ref_input to
+ *          indicate that no fixed reference can be used as input
+ */
+#define ADC_NG_NO_FIXED_INPUT           (UINT8_MAX)
+
+/**
+ * @brief   Flag to indicate the MCUs supply voltage is used as reference
+ *
+ * When a known, lower reference voltage can be selected as input and is
+ * sampled using the MCU's VCC as reference, the MCU's VCC can be deduced.
+ */
+#define ADC_NG_REF_MCU_VCC              (0x8000U)
+
+/**
+ * @brief   Flag a reference voltage as calibrated
+ *
+ * E.g. the 1.1V bandgap voltage reference available on many ATmega has
+ * production variations of +-100mV, but is extremely stable; so it will
+ * consistently create the same voltage for a wide range of supply voltage and
+ * operating temperatures. So if the actual value of the reference is e.g.
+ * written to EEPROM and restored on boot, the reference value should be marked
+ * as calibrated.
+ */
+#define ADC_NG_REF_CALIBRATED           (0x4000U)
+/**
+ * @brief   Access the voltage value without flags
+ */
+#define ADC_NG_REF_MASK                 (0x3fffU)
+
+/**
+ * @brief   Description of an thermistor to use for temperature measurements
+ */
+typedef struct {
+    /**
+     * @brief   Contains the temperature coefficient of the NTC, or zero
+     *
+     * The coefficient is is given in 1/1024 mV per 0.1 °C. The resulting
+     * temperature in 0.1 °C is calculated from the measured voltage using:
+     *
+     *    T[d°C] = (coefficient * (mV - offset)) / 1024
+     */
+    uint16_t coefficient;
+    /**
+     * @brief   The offset in mV to use for obtaining the temperature
+     *
+     * See @ref adc_ng_ntc_t::coefficient
+     */
+    uint16_t offset;
+} adc_ng_ntc_t;
+
+/**
+ * @brief   Internal driver interface
+ */
+typedef struct {
+    /**
+     * @brief   Initialize the given ADC channel and prepare it for sampling
+     * @param           handle  Handle of the ADC
+     * @param[in]       chan    The ADC channel to initialize
+     * @param[in]       res     The resolution to select
+     * @param[in        ref     Index of the reference to use (@ref adc_ng_driver_t::refs)
+     *
+     * @retval  0               Success
+     * @retval  -ENXIO          Invalid channel given in @p channel
+     * @retval  -EALREADY       The ADC is already powered and configured
+     * @retval  <0              Other error (see device driver doc)
+     *
+     * @post    When 0 is returned, the channel @p channel is ready take samples
+     * @post    If @p res contains an unsupported resolution, an assert blows up
+     * @post    If `-EALREADY` is returned, the ADC is state remains unchanged
+     * @post    On other error codes the ADC is powered down
+     *
+     * @note    A call to @ref adc_ng_driver_t::off is needed to disable the channel
+     *          again and preserve power
+     */
+    int (*init)(void *handle, uint8_t chan, uint8_t res, uint8_t ref);
```

Comment:
It might be useful to define these driver functions as separate typedefs, e.g. `adc_ng_init_t` for the following two reasons:
1. Simplifies casting driver function that might be necessary when setting the driver from
    ```c
    .init = (int (*)(void *handle, uint8_t chan, uint8_t res, uint8_t ref))foo_adc_ng_init;
    .off = (void (*ff)(void *handle))foo_adc_ng_init;
    ...
    ```
    to
   ```c
    .init = (adc_ng_init_t)foo_adc_ng_init;
    .off = (adc_ng_off_t)foo_adc_ng_off;
    ```    
2. `adc_ng_*_t` could be referenced in the documentation of the driver.

Reply:
Agreed! Will do so tomorrow

Reply:
Done :-)